src/timesync/timesyncd.c

```diff
@@ -91,49 +88,47 @@ static int load_clock_timestamp(uid_t uid, gid_t gid) {
                 if (r < 0)
                         log_debug_errno(r, "Failed to create "TIMESYNCD_CLOCK_FILE_DIR", ignoring: %m");
 
-                /* create stamp file with the compiled-in date */
-                r = touch_file(TIMESYNCD_CLOCK_FILE, /* parents= */ false, min, uid, gid, 0644);
+                /* Create stamp file with the compiled-in date */
+                r = touch_file(TIMESYNCD_CLOCK_FILE, /* parents= */ false, epoch, uid, gid, 0644);
                 if (r < 0)
                         log_debug_errno(r, "Failed to create %s, ignoring: %m", TIMESYNCD_CLOCK_FILE);
         } else {
                 struct stat st;
-                usec_t stamp;
 
-                /* check if the recorded time is later than the compiled-in one */
+                /* Check if the recorded time is later than the compiled-in one */
                 if (fstat(fd, &st) < 0)
                         return log_error_errno(errno, "Unable to stat timestamp file "TIMESYNCD_CLOCK_FILE": %m");
 
-                stamp = timespec_load(&st.st_mtim);
-                if (stamp > min)
-                        min = stamp;
-
                 /* Try to fix the access mode, so that we can still touch the file after dropping
                  * privileges */
                 r = fchmod_and_chown(fd, 0644, uid, gid);
                 if (r < 0)
                         log_full_errno(ERRNO_IS_PRIVILEGE(r) ? LOG_DEBUG : LOG_WARNING, r,
                                        "Failed to chmod or chown %s, ignoring: %m", TIMESYNCD_CLOCK_FILE);
 
-                (void) advance_tstamp(fd, &st);
+                epoch = MAX(epoch, timespec_load(&st.st_mtim));
+
+                (void) advance_tstamp(fd, epoch);
         }
 
         ct = now(CLOCK_REALTIME);
-        if (ct > min)
+        if (ct > epoch)
                 return 0;
 
         /* Not that it matters much, but we actually restore the clock to n+1 here rather than n, simply
          * because we read n as time previously already and we want to progress here, i.e. not report the
          * same time again. */
-        if (clock_settime(CLOCK_REALTIME, TIMESPEC_STORE(min+1)) < 0) {
-                log_warning_errno(errno, "Failed to restore system clock, ignoring: %m");
+        if (clock_settime(CLOCK_REALTIME, TIMESPEC_STORE(epoch + 1)) < 0) {
+                log_warning_errno(errno, "Failed to advance system clock, ignoring: %m");
                 return 0;
         }
 
         log_struct(LOG_INFO,
                    "MESSAGE_ID=" SD_MESSAGE_TIME_BUMP_STR,
-                   "REALTIME_USEC=" USEC_FMT, min+1,
-                   LOG_MESSAGE("System clock time unset or jumped backwards, restored from recorded timestamp: %s",
-                               FORMAT_TIMESTAMP(min+1)));
+                   "REALTIME_USEC=" USEC_FMT, epoch + 1,
+                   LOG_MESSAGE("System clock time advanced to %s: %s",
```

Comment:
Please update the corresponding catalog entry. Plus, I think we should use the structured message in pid1 too.

Reply:
Actually, the catalog entry is already correct. timesyncd only advances the clock based on a timestamp. It's systemd that has more complicated logic.